### DISCUSSION TITLE
Discussion: Allow extension classes to be nested

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#5803] [Language Ideas]
```cs
class C {
  void M(object o) => o.Chainable().Chainable();
  static class PrivateExtentions {
    // can capture outer type parameters
    // can access outer private members
    // takes precedence over outer extensions
    // only visible from the containing type
    public static object Chainable(this object o) { .. }
  }
}
```

**Comment by MillKaDe** [comment]
I would like to suggest a slightly different syntax for extension functions with reduced accessibility:

```c#
class C {
  private static object Chain1 (this object o) { ... } // accessible from C and types nested inside C
  protected static object Chain2 (this object o) { ... } // accessible from C and types nested inside C and types derived from C
  void M (object o) => o.Chain1 ().Chain1 (); // ok; same as Chain1 (Chain1 (o));
  class N { // nested in C
    void MN1 (object o) => o.Chain1 ().Chain1 (); // ok; same as Chain1 (Chain1 (o));
    void MN2 (object o) => o.Chain2 ().Chain2 (); // ok; same as Chain2 (Chain2 (o));
  } // class N
} // class C
class D : C { // derived from C
  void MD1 (object o) => o.Chain1 ().Chain1 (); // error: Chain1 not accessible, cause D is not nested in C
  void MD2 (object o) => o.Chain2 ().Chain2 (); // ok; same as Chain2 (Chain2 (o));
} // class D
class E { // elsewhere ..
  void ME1 (object o) => o.Chain1 ().Chain1 (); // error: Chain1 not accessible, cause E is not nested in C
  void ME2 (object o) => o.Chain2 ().Chain2 (); // error: Chain2 not accessible, cause E is not derived from C
} // class E
```
The nested static class PrivateExtensions is not really needed, because the extension funcs could be defined as static members of the class/type that would contain class PrivateExtensions.

The (only) advantage of extension funcs with reduced accessibility would be, that they can be called as if they were members of their first parameter ...
```c#
class C {
  static object X (this object o) => { ... }
  static object Y (this object o) => { ... }
  static object Z (this object o) => { ... }
  void MStatic (object o) => Z (Y (X (o))); // use extension funcs X,Y,Z by calling them as static funcs
  void MExt (object o) => o.X ().Y ().Z (); // use extension funcs X,Y,Z by calling them as extension funcs
} // class C  
```
... which sometimes might be more readable because of the (non-)inverted sequence.

Questions:
- what is the default, if the private / protected modifier is omitted ? I would suggest that it means private.
- what would a public modifier mean ? I would suggest it should be treated as an error, unless the containing class can be a extension class (because it is static and public).

**Comment by bernd5** [comment]
Why do you want to use an extensions method in such a case instead of a regular method?

**Reply by MillKaDe** [reply]
> Why do you want to use an extensions method in such a case instead of a regular method?

Extensions functions are just syntactic sugar to call static functions as if they were instance members of their first parameter.

For example, consider this simple LINQ queries :
```c#
using System;
using System.Linq;
using E = System.Linq.Enumerable;
public static class Program {
  static bool IsOdd (int i) => 0 != (i & 1);
  static int ME () => E.Range (1, 10).Where (IsOdd).Last ();
  static int MS () => E.Last (E.Where (E.Range (1, 10), IsOdd));
  static void Main () {
    var me = ME ();
    var ms = MS ();
  }
}
```
.. and check at SharpLab.IO
- the [generated IL code](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyAMABMugHQAyYAdgI4DcO+BAogQLxGkU0lOUCuAWwCmAJwCGAIwhD6uVAGYCAbxwEOANgISA9togEAkgGcA8jBgEAFFWAEwASjYA+AoQCE7awQBkBdPdk1Yk0bAgBZFktHVhcmEgAlMUoAcyErdCg/PHsSAHUAC1E0y2MzGByyMSNbKMCNO0pbMIBlK2jY8iqauIKiqzjElOKMrPtM0vN7ANV65AAWcLEqNuUZtQA3MRECYTZwyOnsNQ2tnaM9lra6gF8ca6A) as well as
- the [decompiled C# code](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQBkCWAdgI4Dcya6AougLyY4EnZWECuAtgKYBOAhgCMANl3IoYAZnQBvZOgYA2dAID2KoegCSAZwDywYOgAURAC7p8ASjoA+dBgCE9E+gBk6OJbHysSs+gBZGiNrWjsqbAAlPkIAcy5jOAAaD1RLbAB1AAteBKMdfWB03D5tcxDvRQtCcwCAZWNQ8LxS8ojs3OMI6Li85NTLFIKDSy85KqgAFkC+IkaZcfkANz4edG46QOCxpHll1fXtTfrGyoBfZDOgA==) .

Both functions ME and MS generate exactly the same IL code.  The extension syntax version (ME) gets 'lowered' by the compiler to the static syntax version (MS).  **This lowering requires to invert the call sequence and convert chaining to nesting.**

Now consider a real world query with three linked database tables (Customer, Order, Item), several filter predicates for each table, and then some other clauses for ordering, left and right join and whatever else.

Such a query would still be readable with extension syntax, but its static syntax form would look quite ugly.
